examples/pytorch/hilander/utils/knn.py

```diff
@@ -0,0 +1,173 @@
+        sims.fill(fill_value)
+    row, col = np.where(sims >= th_sim)
+    # remove the self-loop
+    idxs = np.where(row != nbrs[row, col])
+    row = row[idxs]
+    col = col[idxs]
+    data = sims[row, col]
+    col = nbrs[row, col]  # convert to absolute column
+    assert len(row) == len(col) == len(data)
+    spmat = csr_matrix((data, (row, col)), shape=(n, n))
+    return spmat
+
+def build_knns(feats,
+               k,
+               knn_method,
+               dump=True):
+    with Timer('build index'):
+        if knn_method == 'faiss':
+            index = knn_faiss(feats, k, omp_num_threads=None)
+        elif knn_method == 'faiss_gpu':
+            index = knn_faiss_gpu(feats, k)
+        else:
+            raise KeyError(
+                'Only support faiss and faiss_gpu currently ({}).'.format(knn_method))
+        knns = index.get_knns()
+    return knns
+
+
+class knn():
+    def __init__(self, feats, k, index_path='', verbose=True):
+        pass
+
+    def filter_by_th(self, i):
+        th_nbrs = []
+        th_dists = []
+        nbrs, dists = self.knns[i]
+        for n, dist in zip(nbrs, dists):
+            if 1 - dist < self.th:
+                continue
+            th_nbrs.append(n)
+            th_dists.append(dist)
+        th_nbrs = np.array(th_nbrs)
+        th_dists = np.array(th_dists)
+        return (th_nbrs, th_dists)
+
+    def get_knns(self, th=None):
+        if th is None or th <= 0.:
+            return self.knns
+        # TODO: optimize the filtering process by numpy
+        # nproc = mp.cpu_count()
+        nproc = 1
+        with Timer('filter edges by th {} (CPU={})'.format(th, nproc),
+                   self.verbose):
+            self.th = th
+            self.th_knns = []
+            tot = len(self.knns)
+            if nproc > 1:
+                pool = mp.Pool(nproc)
+                th_knns = list(
+                    tqdm(pool.imap(self.filter_by_th, range(tot)), total=tot))
+                pool.close()
+            else:
+                th_knns = [self.filter_by_th(i) for i in range(tot)]
+            return th_knns
+
+class knn_faiss(knn):
+    def __init__(self,
+                 feats,
+                 k,
+                 nprobe=128,
+                 omp_num_threads=None,
+                 rebuild_index=True,
+                 verbose=True,
+                 **kwargs):
+        import faiss
+        if omp_num_threads is not None:
+            faiss.omp_set_num_threads(omp_num_threads)
+        self.verbose = verbose
+        with Timer('[faiss] build index', verbose):
+            feats = feats.astype('float32')
+            size, dim = feats.shape
+            index = faiss.IndexFlatIP(dim)
+            index.add(feats)
+        with Timer('[faiss] query topk {}'.format(k), verbose):
+            sims, nbrs = index.search(feats, k=k)
+            self.knns = [(np.array(nbr, dtype=np.int32),
+                          1 - np.array(sim, dtype=np.float32))
```

Comment:
One caveat here is that we would need to add a clamping of the cosine similarity to avoid the assertion error on line 55. No clamping should be ok most of the times but there are occasionally some nodes with neighboring large distances will cause this issue.

`                self.knns = [(np.array(nbr, dtype=np.int32),
                    1.0 - np.clip(np.array(sim, dtype=np.float32), 0.0, 1.0))
                    for nbr, sim in zip(nbrs, sims)]`

Reply:
Discussed offline, it's OK to release without them since we don't need that to reproduce the numbers in the paper. If we verify no number change when adding this line or we need to add some numbers for dataset requires this line, let's add that.